src/connectors/crowdstrike_devices.py

```diff
@@ -256,8 +308,258 @@ def ingest(table_name, options):
                 )
                 for device in devices
             ],
-            select=db.derive_insert_select(LANDING_TABLE_COLUMNS),
-            columns=db.derive_insert_columns(LANDING_TABLE_COLUMNS),
+            select=db.derive_insert_select(LANDING_DEVICES_TABLE_COLUMNS),
+            columns=db.derive_insert_columns(LANDING_DEVICES_TABLE_COLUMNS),
         )
         log.info(f'Inserted {len(devices)} rows.')
         yield len(devices)
+
+
+rem_id_cache = {}
+
+
+async def does_rem_exist(rem_id: str) -> str:
+    """
+    does_rem_exist returns "" if the remediation is has already been cached locally
+    or is recorded in the database. Otherwise, it returns the remediation ID.
+    """
+    # Check local cache if you have seen this ID in this run
+    if rem_id in rem_id_cache:
+        log.info(f"[CACHE]: Local cache hit! Already recorded {rem_id} in this round")
+        return ""
+
+    # Check if we have it stored in the table already
+    row = db.fetch_latest(
+        "data.CROWDSTRIKE_SPOTLIGHT_REMS_DEFAULT_CONNECTION",
+        where=f"REM_ID='{rem_id}'",
+    )
+    if row is not None:
+        log.info(f"[CACHE]: Database cache hit! Already recorded {rem_id} in Snowflake")
+        rem_id_cache[rem_id] = True
+        return ""
+
+    return rem_id
+
+
+async def get_uncached_rem_dets(vuln_dets: list) -> Coroutine[None, None, list]:
+    """
+    get_uncached_rem_dets returns the remediation details of only remediations that have not been recorded yet.
+    According to the documentation, it pretty common for many of the vulns to share the same remediations.
+    """
+    # dedupe all the remediation IDs from all the vuln detail information.
+    rem_ids = set(
+        reduce(
+            (lambda x, y: x + y if (y != None) else x),
+            [det.get("remediation", {}).get("ids", []) for det in vuln_dets],
+            [],
+        )
+    )
+    # find the IDs that we have not already recorded
+    rem_ids = await asyncio.gather(
+        *[asyncio.create_task(does_rem_exist(r)) for r in rem_ids]
+    )
```

Comment:
This looks like an "n+1 query", given that you have the whole set of rem_ids would it be possible to issue a query to look for the whole set in a batch?

Reply:
Yup, good call. Fixed this